### ISSUE TITLE
Support repeated request headers using HTTPHeaderDict

### ISSUE THREAD

**Issue by jkupferer** [urllib3/urllib3#2242]
### Context

What are you trying to do?

I am trying to pass repeated headers such as the `Impersonate-Group` header used by the kubernetes API. (https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation)

How do you expect to do it?

I would like to be able to call the urllib3 request method and pass the parameter `headers` as an HTTPHeaderDict.

Is it something you currently cannot do?

It is nearly supported. It is broken for requests with request body because of the implementation of `request_encode_body`:

https://github.com/urllib3/urllib3/blob/main/src/urllib3/request.py#L174
```
        extra_kw: Dict[str, Any] = {"headers": {}}
```

https://github.com/urllib3/urllib3/blob/main/src/urllib3/request.py#L196
```
        extra_kw["headers"].update(headers)
```

These lines force headers to be a plain python dictionary even when an HTTPHeaderDict is passed.

Is this related to an existing issue/problem?

Not that I am aware of.

### Alternatives

Can you achieve the same result doing it in an alternative way?

Not that I know of

Is the alternative considerable?

### Duplicate

Has the feature been requested before?
If so, please provide a link to the issue.

Not that I know of.

### Contribution

Would you be willing to submit a PR?
_(Help can be provided if you need assistance submitting a PR)_

Yes

**Comment by shazow** [comment]
+1, I believe this used to be a feature we supported at one point! Specifically, you could pass in a list of tuples which `dict.update(...)` accepted.

If the only thing blocking this is the new type annotations, makes sense to fix that. Let's wait for Seth/Quentin to give guidance on the best way to move forward though. :)

**Comment by sethmlarson** [comment]
So `HTTPHeaderDict` for requests is an interesting one, because it makes total sense to use but it really wasn't meant to be a public API in v1.x. Users are still able to import it from `urllib3.connection` and `urllib3.response` and use it without knowing it isn't a public API though. But now when you try to pass that `HTTPHeaderDict` as request headers in v1.x it... almost works? And that's an unfortunate thing for us cuz now our users are using something that isn't tested anywhere!

So starting in v2.0 we're making `HTTPHeaderDict` a first-class citizen and actually the recommended way of managing headers for both requests and responses.

Basically all of this is to say we don't support duplicate headers with any released version, sorry! My recommendation is to try using commas to combine the header values. You can also use an `HTTPHeaderDict` to do this for you like so:

```python
header_dict = HTTPHeaderDict()
header_dict.add("Header", "1")
header_dict.add("Header", "2")
headers = dict(header_dict.itermerged())
{"Header": "1, 2"}
```

(Btw the `main` branch that you linked to is for v2 development and hasn't been released yet, we have a `1.26.x` branch for `v1.x` releases)

**Comment by pquentin** [comment]
(Are we sure this works correctly in the main branch? We didn't add tests when documenting it as supported.)

**Comment by sethmlarson** [comment]
@pquentin Doesn't work in the main branch, just signaled the intent but should probably also create an issue to track it too :)

**Comment by sethmlarson** [comment]
For now going to use this issue as it contains all the discussion, I'll rename and add to the v2 milestone.

**Comment by jkupferer** [comment]
Thank you for your attention on this issue.
@sethmlarson I tried using commas in my values, but unfortunately the kubernetes API (and many others it seems) do not follow RFC 7230 https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.2

"A sender MUST NOT generate multiple header fields with the same field name in a message unless either the entire field value for that header field is defined as a comma-separated list [i.e., #(values)] or the header field is a well-known exception (as noted below)."

So it seems to me the current behavior of urllib3 is technically correct, but unfortunately prevents communicating with some apis.

**Comment by pquentin** [comment]
I've discussed the issue of allowing multiple headers with a few maintainers. There's no question that this is something we want to support, given the real-words use cases that need it.

We currently have two proposals, both handling overwrite, sending multiple headers, and folding. Suppose we want to send those headers:

```
Impersonate-User: jane.doe@example.com
Impersonate-Group: developers
Impersonate-Group: admins
Cache-Control: public, max-age=604800
```

It seems to me that neither of those proposals is breaking, so this change does not have to be in 2.0?

## First proposal: `add(fold: bool)`

```python3
headers = HTTPHeaderDict()

headers["Impersonate-User"] = "john.doe@example.com"
headers["Impersonate-User"] = "jane.doe@example.com"

headers.add("Cache-Control", "public")
headers.add("Cache-Control", "max-age=604800")

headers.add("Impersonate-Group", "developers", fold=False)
headers.add("Impersonate-Group", "admins", fold=False)
```

## Second proposal: `add(on_duplicate=overwrite|fold|sendmultiple)`

(Using strings in this example but it would probably be an enum instead)

```python3
headers = HTTPHeaderDict()

# we now have two ways to overwrite
headers["Impersonate-User"] = "john.doe@example.com"
headers.add("Impersonate-User",  "martin.doe@example.com", on_duplicate="overwrite")
headers["Impersonate-User"] = "jane.doe@example.com"

# on_duplicate=fold is the default behavior
headers.add("Cache-Control", "public")
headers.add("Cache-Control", "max-age=604800")

headers.add("Impersonate-Group", "developers", on_duplicate="sendmultiple")
headers.add("Impersonate-Group", "admins", on_duplicate="sendmultiple")
```

**Comment by sigmavirus24** [comment]
So I agree that these are both backwards compatible. I do think @sethmlarson had a good point about `__setitem__` already works to be `add(header_name, header_value, on_duplicate="overwrite")` but here's the thing, I think `__setitem__` could be confusing to have to switch to. I agree, one obvious way to do something is great, and I'm also a fan of "Oh shit, I didn't realize I was doing the wrong thing here, easier to add a parameter to do the right behaviour and iterate until I find the right output". 

Part of me wants a "fourth" option if we were to go with `on_duplicate` as well but I don't know that it matters for `HTTPHeaderDict` because the primary use for that case is as a server, not so much a client. I'll explain it anyway as food for thought.

Imagine that the default for `on_duplicate` is `auto`. That means for a sub-set of headers that we know are absolutely safe to send as multiple and should prefer not to be sent as a folded header (`Set-Cookie` is the most common one of these). For other headers `fold` is the right "automatic" behaviour.

If our header dict were used more widely, I'd be arguing for `auto` but I'm too sleep deprived to think of a client case that's similar that would require us to be smart for the user.

**Comment by pquentin** [comment]
@haikuginger @shazow What are your thoughts on those APIs?

**Comment by shazow** [comment]
One suggestion I would consider is changing the term "fold" to "combined", since that's the term RFC 7230 uses and I think "fold" is not very beginner-programmer friendly (more of a functional programming term?).

If I was picking terms from scratch, my gut is to go with "replace", "combine", "separate".

If we wanted to go with UPSERT terminology, it's ignore/insert/update/merge. (Worth adding a 4th `ignore` option? What about a 5th `unique` option that throws?)

Other options I can think of that aren't great:
- Could have more `add`-like methods which have different names, such as `add_combined` or something like `append` or `insert`.
- Could have different HTTPHeaderDict implementations with different behaviours. (Is it easy to bring your own implementation rn or are we forcing people to use ours?)

In the spirit of making progress on this, my vote would be doing either of the current `add` proposals and renaming `fold` to `combine`. :)

**Comment by sethmlarson** [comment]
Using `combine=False` instead of `fold=False` works for me too. This is the verb that RFC 7230 uses so is consistent there too. Great suggestion, @shazow!

**Comment by haikuginger** [comment]
Haha, ended up going down a rabbit hole on this one trying to find an interface that suits the more complicated cases that `Dict` and `List[Tuple[str, str]]` are insufficient for. If I were doing it from scratch (which we can think about for 2.0, because we're breaking a lot of stuff!), I might consider something like this:

```python3
class SingleValueHeader:
    value: str

    def try_add_value(self, value: str) -> bool:
        return False

    def serialized(self) -> str:
        return self.value

class MultiValueHeader:
    values: List[str]

    def try_add_value(self, value: str) -> bool:
        self.values.append(value)
        return True

    def serialized(self) -> str:
        return ','.join(self.values)


class HttpHeader(Protocol):
    
    def try_add_value(self, value: str) -> bool:
        """
        Returns True if the value was successfully merged into this header;
        returns False if this header cannot accept additional values
        """

    def serialize(self) -> str:
        """
        Return the string that this header should be serialized to in a request.
        """

class RequestHeaders:

    def all_by_name(self, name: str) -> List[HttpHeader]:
        """
        Return an iterable of all the `HttpHeader` entities that will be
        serialized and sent over the wire when this request is placed
        """

    def set_header(self, name: str, header: Union[str, HttpHeader]) -> None:
        """
        Ensures there is exactly one header with the provided name, and it has
        exactly the single serializable value provided (whether that value is
        a single string or a collection of strings that will be concatenated
        together when placing the request)
        """

    def serializable_kv_pairs(self) -> List[Tuple[str, str]]:
        """
        Return a list of string-string pairs that can be safely transmitted
        over the wire as part of an HTTP request
        """

    @overload  # When an HttpHeader object is passed, `mergeable` is disregarded
    def add_header(self, name: str, header: HttpHeader) -> None: ...
    @overload
    def add_header(self, name: str, header: str, mergeable: bool = False) -> None: ...
    @overload
    def add_header(
        self,
        name: str,
        header: Union[str, HttpHeader],
        mergeable: bool = False
    ) -> None:
        """
        In addition to any other headers that will be included in this request,
        send this header with the value specified; send it in addition to any
        headers that bear the same name but will be sent separately.

        If `mergeable` is True, the header will be created in such a way that
        `add_header_value` may add additional contents to it later on, but only
        if the value passed is a plain string; if it's `HttpHeader`, then trust
        it to handle its own business when `try_add_value` is called.
        """

    def add_header_value(self, name: str, value: str, mergeable: bool = False) -> None:
        """
        Ensure that the provided value is represented in the set of headers
        that will be serialized for this request.
        """

        if mergeable:
            # if this value can share a header with other values, get any existing
            # headers we might add this value to
            potential_existing_headers = self.all_by_name(name)

            # iterate over those headers...
            for header in potential_existing_headers:
                # ...trying to add the value to each in turn...
                if header.try_add_value(value):
                    # and finishing if we did it!
                    return

        # if we didn't find a header we could add this value to, or if it can't
        # be merged with others, create a new header that can accommodate it,
        # making sure that the new header follows this value's merge rules.
        return self.add_header(name, value, mergeable=mergeable)
```

I think the trick here is that for the case where we're adding values, we want to make sure that each value for a given name agrees with other already-added values about whether they can be merged together. That means that for the complex cases, we need to carry around a bit more than just strings.

**Comment by rtpg** [comment]
I tried to the API described [in this comment](https://github.com/urllib3/urllib3/issues/2242#issuecomment-1023764816) in the PR #2652. 

The biggest messiness comes from backwards compatibility for the object. In particular, around the `items` method. I ended up with just one toggle to get this work, but we could go for zero if we are OK with the break on that specific method.

I would appreciate reviews on that PR to see if that makes sense.
